### PR TITLE
Drop first class support for 4.07

### PR DESCRIPTION
**Pull request by ghost**
Following the release of OCaml 4.10, this PR moves the minimum version of OCaml required to build Dune itself to 4.08. In particular, it deletes the src/ocaml-syntax-shims directory which is no longer needed.

TODO:
- [x] wait for a 4.10 compatible merlin
- [x] wait for a 4.10 compatible odoc
- [ ] wait until all the Dune devs are happy to use 4.10 to work on Dune
- [x] delete all the `(preprocess future_syntax)` from the source tree

NEXT:
- enjoy all the 4.08 features!

### CODE REVIEW

**Comment by ejgallego** [comment]
Hi @diml maybe this issue is not the best place to discuss compat [please redirect me to the proper place] but there is a small problem with the way deps are being bumped:

Some packages do depend on `dune-configurator`, which depends on `dune-private-libs` which has a lower bound of 4.08.0 in this PR. This means that once Dune 2.4.0 is out none of these packages will be installable with OCaml < 4.08.

An example is `cairo2` which means that lablgtk3 indeed cannot be installed in 4.06 for example as of today unless you pin Dune to 2.0.0 [we have to do this on Coq's CI] I am not sure the dependency on those packages is the right thing to do BTW.

Anyways, the current situation seems a bit restrictive to me; for example this PR would force Coq's minimal version to 4.08 which indeed has some other serious unsolved problems for Coq users.

**Comment by diml** [comment]
For configurator, as long you don't rely on recently added features you should still be installable with older OCaml versions. For instance, if all you need is in `dune-configurator` 2.0.0, then you can depend on `dune-configurator >= 2.0.0`. When using an older compiler, you could install an older configurator and when using a newer one you would have to install a recent configurator. That seems acceptable to me, what do you think?

For Coq, would the scheme we are using for opam work as well? i.e. build dune with a recent OCaml and then use it to build projects with an older OCaml?

In any case, there is no rush. So if this bump is an issue for Coq, let's wait until we find a satisfactory solution. That's why I opened this PR early :) So that we can discuss and find a way forward.

**Comment by ejgallego** [comment]
> When using an older compiler, you could install an older configurator and when using a newer one you would have to install a recent configurator. That seems acceptable to me, what do you think?

That seems a bit restrictive tho; for example, most users that'd like to use Dune 2.3 are effectively stuck with OCaml >= 4.07 , despite the `ocaml-secondary-compiler` trick, as `dune-private-libs` forces OCaml >= 4.07 anyways.

> For Coq, would the scheme we are using for opam work as well? i.e. build dune with a recent OCaml and then use it to build projects with an older OCaml?

That doesn't work due to the deps of dune-private-libs

> So if this bump is an issue for Coq, let's wait until we find a satisfactory solution. That's why I opened this PR early :) So that we can discuss and find a way forward.

Indeed thanks; IMO this is a more general issue than Coq, as it affects the whole package repos.

[For Coq in particular would be pretty problematic tho as it would mean that we force 4.08 on users, which we cannot do, but IMO we shouldn't base the policy on what Coq needs]

**Comment by avsm** [comment]
I don't think that using the opam solver for compatibility with the dune libraries is going to work that well.

As an example, dune 2.3 changed the interface between configurator and dune, and so trying to build duniverse with a newer dune (>=2.3) but older dune-configurator (<2.3) breaks.  You can see this in the wild in the duniverse tool itself (https://github.com/ocamllabs/duniverse/issues/77#issuecomment-591968453).

Perhaps we need to explicitly version the interface between the host `dune` binary and its libraries. The interfaces don't change that often, but when they do, it is essential to have a matching `dune` binary.

**Comment by diml** [comment]
Yh, the old communication protocol between dune and configurator wasn't great. It relied on configurator executing `ocamlc -config` on every single run. The new one is simpler and there is much less of a need to break things.

Configurator was one of the last component that wasn't versioned. It now has:

```
(library
 (name configurator)
 (public_name dune-configurator)
 ...
 (special_builtin_support
  (configurator
   (api_version 1))))
```

The important bit is the last line. We don't have all the infrastructure in place to support multiple versions simultaneously, but at least it'll be possible to upgrade the protocol in the future without breaking things.

**Comment by diml** [comment]
> That seems a bit restrictive tho; for example, most users that'd like to use Dune 2.3 are effectively stuck with OCaml >= 4.07 , despite the ocaml-secondary-compiler trick, as dune-private-libs forces OCaml >= 4.07 anyways.

Yes, but you should be able to use `dune-private-libs` 2.0.1 with dune 2.3.0 and a 4.07 compiler for instance. Do you agree?

**Comment by ejgallego** [comment]
> Yes, but you should be able to use `dune-private-libs` 2.0.1 with dune 2.3.0 and a 4.07 compiler for instance. Do you agree?

Indeed, ok if the deps on configurator / private libs are going to be kept loose like that things could work.

**Comment by diml** [comment]
Yh, the idea is that the various `dune-xxx` packages have very tight dependency constraints between them. So dune-configutor x.y depends on dune-private-libs x.y exactly. This is for simplicity, in particular we don't want to commit to stability for the various libraries in `dune-private-libs`. However, they are all normal OCaml packages and all depend on `dune >= x.y`. So they should all be buildable with future dune releases.

The only case that doesn't work is using a feature from a recent dune-configurator with an older compiler.

So essentially as a user you have to choose between using recent features of dune-configurator vs being compatible with older OCaml compiler. Which seems like an acceptable compromise to me. In particular, `dune-configurator` doesn't change that much anymore.

So in the end, is it fine to proceed with this PR from Coq point of view? Once 4.10 is ready to be used for development of course.

**Comment by ejgallego** [comment]
> So in the end, is it fine to proceed with this PR from Coq point of view?

It should be fine, thanks @diml

File: src/dune/ocaml_version.ml [hunk 1]
```diff
@@ -32,7 +32,7 @@ let ooi_supports_no_code version = version >= (4, 05, 0)
 
 let supports_let_syntax version = version >= (4, 08, 0)
 
-let supports_output_complete_exe version = version >= (4, 10, 0)
+let supports_output_complete_exe version = version >= (4, 10, 1)
```
Comment (jeremiedimino) [inline src/dune/ocaml_version.ml:35]: /cc @nojb. I bumped to 4.10.1 because of ocaml/ocaml#9495

Comment (nojb) [reply src/dune/ocaml_version.ml:35]: Yes, looks good.

**Comment by jeremiedimino** [comment]
github is being difficult, I'm re-opening a new PR to check CI after rebase

**Pull request closed without merging**